drivers/i3c/i3c_dw.c

```diff
@@ -0,0 +1,1743 @@
+	cmd->ccc.data_len = len;
+	cmd->targets.payloads->rnw = (rnw ? 1 : 0);
+	cmd->ccc.id = id;
+}
+
+/**
+ * @brief Send CCC command to enable/disable events
+ *
+ * @param master: Pointer to the master device
+ * @param addr: Address of the slave device
+ * @param evts: Events
+ * @param enable: Flag to enable/disable
+ *
+ * @retval 0 if successful or error code
+ */
+static int32_t i3c_master_en_dis_evt(const struct device *dev,
+				     uint8_t addr, uint8_t evts, bool enable)
+{
+	struct i3c_ccc_payload cmd;
+	struct i3c_ccc_events events;
+	uint8_t id = (enable ? I3C_CCC_ENEC(addr == I3C_BROADCAST_ADDR) :
+			I3C_CCC_DISEC(addr == I3C_BROADCAST_ADDR));
+
+	events.events = evts;
+	i3c_ccc_cmd_init(&cmd, addr, &events, sizeof(events), 0, id);
+
+	return dw_i3c_send_ccc_cmd(dev, &cmd);
+}
+
+/**
+ * @brief Enable the In-Band Interrupt (IBI)
+ *
+ * @param dev: Pointer to the device structure
+ * @param dyn_addr: Dynamic address for which IBI to be enabled
+ *
+ * @retval 0 if successful or error code
+ * @retval -EINVAL if fails to get position of assigned address
+ */
+static int32_t dw_i3c_enable_ibi(const struct device *dev,
+				 struct i3c_device_desc *target)
+{
+	const struct dw_i3c_config *config = DEV_CONFIG(dev);
+	uint32_t bitpos, sir_con;
+	int32_t pos;
+
+	pos = get_addr_pos(dev, target->dynamic_addr);
+	if (pos < 0) {
+		LOG_ERR("%s Invalid Slave address\n", __func__);
+		return pos;
+	}
+
+	sir_con = sys_read32(config->regs + IBI_SIR_REQ_REJECT);
+	bitpos = IBI_SIR_REQ_ID(target->dynamic_addr);
+	sir_con &= ~BIT(bitpos);
+	sys_write32(sir_con, config->regs + IBI_SIR_REQ_REJECT);
+
+	i3c_master_en_dis_evt(dev, target->dynamic_addr,
+			      I3C_CCC_EVENT_SIR, true);
+
+	return 0;
+}
+
+/**
+ * @brief Disable the In-Band Interrupt (IBI)
+ *
+ * @param dev: Pointer to the device structure
+ * @param dyn_addr: Dynamic address for which IBI need to disable
+ *
+ * @retval 0 if successful or error code
+ * @retval -EINVAL if fails to get position of assigned address
+ */
+static int32_t dw_i3c_disable_ibi(const struct device *dev,
+				  struct i3c_device_desc *target)
+{
+	const struct dw_i3c_config *config = DEV_CONFIG(dev);
+	uint32_t bitpos, sir_con;
+	int32_t pos;
+
+	pos = get_addr_pos(dev, target->dynamic_addr);
+	if (pos < 0) {
+		LOG_ERR("%s Invalid Slave address\n", __func__);
+		return pos;
+	}
+
+	sir_con = sys_read32(config->regs + IBI_SIR_REQ_REJECT);
+	bitpos = IBI_SIR_REQ_ID(target->dynamic_addr);
+	sir_con |= BIT(bitpos);
+	sys_write32(sir_con, config->regs + IBI_SIR_REQ_REJECT);
+
+	i3c_master_en_dis_evt(dev, target->dynamic_addr,
+			      I3C_CCC_EVENT_SIR, false);
+
+	return 0;
+}
+
+/**
+ * @brief Get address slot status of the bus
+ *
+ * @param addrslots: Pointer to the address slots info
+ * @param addr: Address of the node
+ *
+ * @retval  Address status value if successful
+ * @retval -EINVAL if address slots not initialized
+ */
+int32_t i3c_bus_get_addr_slot_status(uint32_t *addrslots, uint16_t addr)
```

Comment:
Any reason not to use the existing functions in `i3c_common.h`?

....although this whole code skimming through looks eerily familiar to me. I've seen this before years ago when it was written for zephyr v2.3 (I think?). It probably needs to now be considered with what landed in v3.2 a bit more with what's in `i3c_common.h` and `i3c_ccc.h` so there is no duplicate code. Was this not known about before or was there a reason to not use these?

Reply:
No specific reason, I am working on utilizing existing APIs and ensuring it does not break any functionality. Will post updated patch soon.